### PR TITLE
add restarts to nvidia drivers download in cloud-init

### PR DESCRIPTION
**Pull request by jackfrancis**
and only create cloud-init string if necessary

**What this PR does / why we need it**: Adds restarts to apt statements for GPU drivers, only call string concatenator if necessary

- [x] unit tests
- [x] e2e tests

**Release note**:

```
add restarts for GPU driver downloads during provisioning
```

### CODE REVIEW

**Comment by jackfrancis** [comment]
@wbuchwalter Could you sanity check the changes here? Two things:

- we want to add retry to all cloud-init network-dependent operations
- trying to avoid the empty string cloud-init entry, and so am wrapping the Get'er in a conditional func that looks for N series SKUs

**Comment by wbuchwalter** [comment]
@jackfrancis This sounds good to me. However note that the command you are retrying is going to go away with https://github.com/Azure/acs-engine/pull/2219.

**Comment by jackfrancis** [comment]
@wbuchwalter Thanks for looking it over. I'll merger #2219 first and then rebase the approach in this PR on top of the specifics that those changes will bring.

File: pkg/acsengine/engine_test.go [hunk 1]
```diff
@@ -371,3 +371,184 @@ func TestGetGPUDriversInstallScript(t *testing.T) {
 		t.Fatalf("VMSize without GPU should not receive a script, expected empty string, received %v", s)
 	}
 }
+func TestIsNSeriesSKU(t *testing.T) {
+	// VMSize with GPU and NVIDIA agreement for drivers distribution
```
Comment (wbuchwalter) [inline pkg/acsengine/engine_test.go]: nit: should simply read "VMSize with GPU" instead. We only have agreement for NC6, NC12, NC24, NV6, NV12 and NV24 so far.  This is very minor, but I don't want anyone to get confused on this agreement stuff and distributes drivers without NVIDIA approval by mistake.

File: test/e2e/kubernetes/job/job.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+}
+
+// WaitOnReady will call the static method WaitOnReady passing in p.Metadata.Name and p.Metadata.Namespace
+func (j *Job) WaitOnReady(sleep, duration time.Duration) (bool, error) {
+	return WaitOnReady(j.Metadata.Name, j.Metadata.Namespace, sleep, duration)
+}
+
+// Delete will delete a Pod in a given namespace
```
Comment (CecileRobertMichon) [inline test/e2e/kubernetes/job/job.go]: copy-pasta typo should be "Delete will delete a Job ..."

Comment (CecileRobertMichon) [inline test/e2e/kubernetes/job/job.go]: Copy-paste error should be " Delete will delete a Job ... "

File: test/e2e/kubernetes/job/job.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	Status   Status       `json:"status"`
+}
+
+// Status holds job status information
+type Status struct {
+	Active    int `json:"active"`
+	Succeeded int `json:"succeeded"`
+}
+
+// CreateJobFromFile will create a Pod from file with a name
```
Comment (CecileRobertMichon) [inline test/e2e/kubernetes/job/job.go]: Copy paste typo s/Pod/Job

File: test/e2e/kubernetes/job/job.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+// List is a container that holds all jobs returned from doing a kubectl get jobs
+type List struct {
+	Jobs []Job `json:"items"`
+}
+
+// Job is used to parse data from kubectl get jobs
+type Job struct {
+	Metadata pod.Metadata `json:"metadata"`
+	Spec     pod.Spec     `json:"spec"`
```
Comment (CecileRobertMichon) [inline test/e2e/kubernetes/job/job.go]: Just to make sure, pod.Spec referring to pod is intentional right?

File: pkg/acsengine/engine_test.go [hunk 5]
```diff
@@ -371,3 +371,184 @@ func TestGetGPUDriversInstallScript(t *testing.T) {
 		t.Fatalf("VMSize without GPU should not receive a script, expected empty string, received %v", s)
 	}
 }
+func TestIsNSeriesSKU(t *testing.T) {
+	// VMSize with GPU
+	validSkus := []string{
```
Comment (CecileRobertMichon) [inline pkg/acsengine/engine_test.go]: Assuming you got this list from somewhere, I have no idea which are which so I'm gonna trust you on this one

File: test/e2e/kubernetes/job/job.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+)
+
+// List is a container that holds all jobs returned from doing a kubectl get jobs
+type List struct {
+	Jobs []Job `json:"items"`
+}
+
+// Job is used to parse data from kubectl get jobs
+type Job struct {
+	Metadata pod.Metadata `json:"metadata"`
```
Comment (CecileRobertMichon) [inline test/e2e/kubernetes/job/job.go]: this pod

Comment (jackfrancis) [reply test/e2e/kubernetes/job/job.go]: The types are functionally identical in how we currently use them, so I didn't want to create a duplicate type. `job.Spec`, for example, is different than `pod.Spec`, so I created a distinct `job` type definition for it.

**Pull request merged**